### ISSUE TITLE
Pyright fails to assess the event_source decorator from aws_lambda_powertools

### ISSUE THREAD

**Issue by ericbn** [microsoft/pyright#3265]
**Describe the bug**
Pyright fails to assess the event_source decorator from aws_lambda_powertools.

**To Reproduce**
Via command-line (the ❯ represents the command-line prompt):
```
❯ python3 -m venv venv
❯ . venv/bin/activate
❯ pip install aws_lambda_powertools pyright
❯ mkdir src
❯ >>! src/lambda_function.py <<\EOF
import json

from aws_lambda_powertools.utilities.data_classes import event_source, APIGatewayProxyEventV2


@event_source(data_class=APIGatewayProxyEventV2)
def lambda_handler(event: APIGatewayProxyEventV2, context):
    return {
        "statusCode": 200,
        "body": json.dumps({"message": f"hello {event.json_body['name']}!"})
    }
EOF
❯ >>! pyproject.toml <<\EOF
[tool.pyright]
pythonPlatform = "Linux"
pythonVersion = "3.9"
stubPath = ""
executionEnvironments = [
    {root = "src"},
]
EOF
❯ pyright
No configuration file found.
pyproject.toml file found at /tmp/test.
Loading pyproject.toml file at /tmp/test/pyproject.toml
No include entries specified; assuming /tmp/test
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
Searching for source files
Auto-excluding /tmp/test/venv
Found 1 source file
/tmp/test/src/lambda_function.py
  /tmp/test/src/lambda_function.py:6:15 - error: No parameter named "data_class" (reportGeneralTypeIssues)
  /tmp/test/src/lambda_function.py:6:2 - error: Argument missing for parameter "context" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
Completed in 0.767sec
```

**Expected behavior**
No errors should be reported. This works as expected with mypy:
```
❯ mypy src/lambda_function.py
Success: no issues found in 1 source file
```

**Screenshots or Code**
Code used in the steps to reproduce above:
```python
import json

from aws_lambda_powertools.utilities.data_classes import event_source, APIGatewayProxyEventV2


@event_source(data_class=APIGatewayProxyEventV2)
def lambda_handler(event: APIGatewayProxyEventV2, context):
    return {
        "statusCode": 200,
        "body": json.dumps({"message": f"hello {event.json_body['name']}!"})
    }
```

**VS Code extension or command-line**
Using pyright via command-line:
```
❯ pyright --version
pyright 1.1.233
```

**Comment by erictraut** [comment]
The `event_source` decorator in `aws_lambda_powertools` is itself decorated with an unannotated decorator called `lambda_handler_decorator`. Pyright does its best to infer the return type of this decorator. Mypy doesn't implement return type inference, which is why it doesn't generate any errors in this case.

The return type for the `lambda_handler_decorator` decorator function is challenging to infer because it is a polymorphic function and returns different types depending on the parameter input types. It should be annotated with a set of `@overload` signatures. It also makes use of the `functools.wraps` decorator. There's a known bug in the typeshed definition of this decorator. I've submitted a [PR to fix this bug](https://github.com/python/typeshed/pull/6670), but it depends on functionality that some type checkers in the Python ecosystem don't yet support, so it has been blocked while we wait for other type checkers to catch up.

So the real solution to this problem is to get the maintainers of `aws_lambda_powertools` to provide a proper type annotation for the `lambda_handler_decorator` decorator function.

I'll note that `aws_lambda_powertools` is a "py.typed" library (as defined in [PEP 561](https://peps.python.org/pep-0561/)). That means that the library is claiming to contain full type information. The CLI version of pyright has a special option called "--verifytypes" that you can run on a "py.typed" library to validate its "type completeness". Here's the output when I run this tool on `aws_lambda_powertools`.

```
Symbols exported by "aws_lambda_powertools": 2095
  With known type: 1733
  With ambiguous type: 77
  With unknown type: 285

Type completeness score: 82.7%
```

That means about 18% of the symbols exposed by this library are not properly annotated.

We've provided [this guidance](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md) for library maintainers to help them understand what type information expected within a "py.typed" library.

I think pyright is behaving as expecting here, so I'm going to close the issue.

**Comment by ericbn** [comment]
Thanks for the detailed explanation, @erictraut! I'll report this upstream.
